### PR TITLE
8345221: Replace legacy with new Provider APIs in SunNativeGSS

### PR DESCRIPTION
**Pull request by franferrax**
Hi, this pull request modifies the _SunNativeGSS_ security provider to stop using the legacy Provider APIs (`Provider::putAll`) and start using the new ones (`Provider::putService`).

Changes in this pull request have been removed from #15539, as they aren't a dependency for that work, but just an opportunistic enhancement we made there. As a consequence, this pull request is the reverse of 29e9cba970fdbbd2192c8d2de1e93a03b1b0d1e1, the removal of this work from #15539 (except that we later renamed `constructMechMap` as `getMechOIDs`, and `sun.security.action.PutAllAction` was removed by 940aa7c4cf1bf770690660c8bb21fb3ddc5186e4, so it is no longer part of this pull request).

#### Testing

As part of our testing, we observed all the tests pass in the following categories:

* `jdk:tier1` (see [run 12073939529](https://github.com/franferrax/jdk/actions/runs/12073939529) and [run 12129046506](https://github.com/franferrax/jdk/actions/runs/12129046506))
* `jdk/java/security`

Additionally, we found no regressions with respect to the baseline (959fa4a1a35a1bb650ec5888efaf3d0fc8cfb025) in the following category:

* `jdk/sun/security`
    * We obtained the same results in both codebases (the two `<reportDir>/text/summary.txt` files are identical).
      | Results                 |     |
      |-------------------------|----:|
      | Tests that passed       | 797 |
      | Tests that failed       |  16 |
      | Tests that had errors   |   2 |
      | Tests that were not run |  35 |
      | Total                   | 850 |
    * In particular, the tests that enable the _SunNativeGSS_ security provider, all passed:
        * [jdk/sun/security/jgss/GssContextCleanup.java](https://github.com/openjdk/jdk/blob/959fa4a1a35a1bb650ec5888efaf3d0fc8cfb025/test/jdk/sun/security/jgss/GssContextCleanup.java#L48)
        * [jdk/sun/security/jgss/GssNameCleanup.java](https://github.com/openjdk/jdk/blob/959fa4a1a35a1bb650ec5888efaf3d0fc8cfb025/test/jdk/sun/security/jgss/GssNameCleanup.java#L47)
        * [jdk/sun/security/krb5/auto/BasicProc.java](https://github.com/openjdk/jdk/blob/959fa4a1a35a1bb650ec5888efaf3d0fc8cfb025/test/jdk/sun/security/krb5/auto/BasicProc.java#L276)
        * [jdk/sun/security/krb5/auto/Cleaners.java](https://github.com/openjdk/jdk/blob/959fa4a1a35a1bb650ec5888efaf3d0fc8cfb025/test/jdk/sun/security/krb5/auto/Cleaners.java#L176)
        * [jdk/sun/security/krb5/auto/ReplayCacheTestProc.java](https://github.com/openjdk/jdk/blob/959fa4a1a35a1bb650ec5888efaf3d0fc8cfb025/test/jdk/sun/security/krb5/auto/ReplayCacheTestProc.java#L389)
        * [jdk/sun/security/krb5/runNameEquals.sh](https://github.com/openjdk/jdk/blob/959fa4a1a35a1bb650ec5888efaf3d0fc8cfb025/test/jdk/sun/security/krb5/runNameEquals.sh#L99)

This contribution is co-authored by @franferrax and @martinuy.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8345221](https://bugs.openjdk.org/browse/JDK-8345221): Replace legacy with new Provider APIs in SunNativeGSS (**Enhancement** - P4)

### Reviewers
 * [Weijun Wang](https://openjdk.org/census#weijun) (@wangweij - **Reviewer**)

### Contributors
 * Francisco Ferrari Bihurriet `<fferrari@openjdk.org>`
 * Martin Balao `<mbalao@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22461/head:pull/22461` \
`$ git checkout pull/22461`

Update a local copy of the PR: \
`$ git checkout pull/22461` \
`$ git pull https://git.openjdk.org/jdk.git pull/22461/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22461`

View PR using the GUI difftool: \
`$ git pr show -t 22461`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22461.diff">https://git.openjdk.org/jdk/pull/22461.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22461#issuecomment-2507975393)
</details>

### CODE REVIEW

File: src/java.security.jgss/share/classes/sun/security/jgss/wrapper/SunNativeProvider.java [hunk 1]
```diff
@@ -64,10 +63,10 @@ static void debug(String message) {
         System.err.println(NAME + ": " + message);
     }
 
-    private static final HashMap<String, String> MECH_MAP = constructMechMap();
+    private static final Oid[] MECH_OIDS = constructMechMap();
 
     @SuppressWarnings("restricted")
-    private static HashMap<String, String> constructMechMap() {
+    private static Oid[] constructMechMap() {
```
Comment (wangweij) [inline src/java.security.jgss/share/classes/sun/security/jgss/wrapper/SunNativeProvider.java:69]: Maybe change the method name? The return value is not a map now.

**Comment by wangweij** [comment]
Do you need a `noreg-` label for the JBS issue?

File: src/java.security.jgss/share/classes/sun/security/jgss/wrapper/SunNativeProvider.java (hunk 1 above)
Comment (martinuy) [reply src/java.security.jgss/share/classes/sun/security/jgss/wrapper/SunNativeProvider.java]: Sounds good. Changed.

**Comment by martinuy** [comment]
> Do you need a `noreg-` label for the JBS issue?

Sounds good. Added.

**Review by wangweij — APPROVED** [review]

**Pull request closed without merging**